File: 3593_minimum_increments_to_equalize_leaf_paths/Solution.swift
// LeetCode 3593 - Minimum Increments to Equalize Leaf Paths
// https://leetcode.com/problems/minimum-increments-to-equalize-leaf-paths/

class Solution {
    var graph = [[Int]]()
    var cost = [Int]()
    var ans = 0

    func minIncrease(_ n: Int, _ edges: [[Int]], _ cost: [Int]) -> Int {
        graph = Array(repeating: [], count: n)
        for e in edges {
            graph[e[0]].append(e[1])
            graph[e[1]].append(e[0])
        }
        self.cost = cost
        ans = 0
        _ = dfs(0, -1)
        return ans
    }

    func dfs(_ u: Int, _ p: Int) -> Int {
        if graph[u].count == 1 && p != -1 { return cost[u] }
        var childVals = [Int]()
        for v in graph[u] where v != p { childVals.append(dfs(v, u)) }
        if childVals.isEmpty { return cost[u] }
        var mx = 0
        for c in childVals { mx = max(mx, c) }
        for c in childVals where c < mx { ans += 1 }
        return mx + cost[u]
    }
}

File: 3593_minimum_increments_to_equalize_leaf_paths/solution.rs
// LeetCode 3593 - Minimum Increments to Equalize Leaf Paths
// https://leetcode.com/problems/minimum-increments-to-equalize-leaf-paths/

impl Solution {
    pub fn min_increase(n: i32, edges: Vec<Vec<i32>>, cost: Vec<i32>) -> i32 {
        let n = n as usize;
        let mut graph = vec![Vec::<usize>::new(); n];
        for e in &edges {
            graph[e[0] as usize].push(e[1] as usize);
            graph[e[1] as usize].push(e[0] as usize);
        }
        fn dfs(u: usize, p: i32, graph: &[Vec<usize>], cost: &[i32], ans: &mut i32) -> i64 {
            if graph[u].len() == 1 && p != -1 {
                return cost[u] as i64;
            }
            let mut child_vals = Vec::new();
            for &v in &graph[u] {
                if v as i32 == p {
                    continue;
                }
                child_vals.push(dfs(v, u as i32, graph, cost, ans));
            }
            if child_vals.is_empty() {
                return cost[u] as i64;
            }
            let mx = *child_vals.iter().max().unwrap();
            for c in child_vals {
                if c < mx {
                    *ans += 1;
                }
            }
            mx + cost[u] as i64
        }
        let mut ans = 0;
        dfs(0, -1, &graph, &cost, &mut ans);
        ans
    }
}
